Component – Table: Add compare checkboxes

ryelle opened the pull request:
See #94, builds on #387 – This PR adds the ability to enable checkboxes on table rows to filter a report by items. Checking rows, then hitting the Compare button, will update the URL query to update the page. As this is mock data, there isn't a chart, so the page data doesn't update yet.

This works with the Compare filter #364, so changes in one component will update the other.

![screen shot 2018-09-11 at 12 41 33 pm](https://user-images.githubusercontent.com/541093/45374423-0b232100-b5c0-11e8-989b-515ae8a867d0.png)

You can trigger these checkboxes by passing in the `compareBy` prop - this will be the query param used to update the URL. You'll also need to pass in a list of IDs corresponding to each item in a row. When the "Compare" button is clicked, the component updates the URL query to be:

`filter=compare-{ compareBy }&{ compareBy }={ list of selected IDs }`

This should match the parameters in the product filter constants ([example, `product`](https://github.com/woocommerce/wc-admin/blob/master/client/analytics/report/products/constants.js#L31), [example, `product_cat`](https://github.com/woocommerce/wc-admin/blob/master/client/analytics/report/products/constants.js#L54)) – for the product table, we want to compare by product, not product categories, so we use `product`.

**To test**

- Open `client/analytics/report/products/__mocks__/data.js` & update the IDs & names to products in your site
- Open the product report, Analytics > Products
- See the checkboxes, select a few
- Click the Compare button in the table header – the Compare filter should appear, and fill in with your selected products as tags (this won't work if you don't update the mock data above)
- Reload the page, the selected products should still fill in the Compare filter, and be selected in the table
- Update the Compare filter by searching for a product in the table, click the filter card's Compare button
- The corresponding row in the product table should appear checked
- Clear all by Xing the tags or unchecking the rows, make sure the two stay synced
- Check that the Revenue report is not updated by this change

Note: Switching from "Show: Product Comparison" to "Show: All Products" does not remove the products query right now, so that _does not_ currently clear the selected rows.

ryelle committed:
Switch to withInstanceId higher order component

ryelle committed:
Add checkboxes to the rows in a TableCard, if a compareBy prop is set

ryelle committed:
Add Compare button to update query param

ryelle committed:
Populate the selected rows from the query parameter

ryelle committed:
Update compare filter display & selected table rows when the query changes

ryelle committed:
Skip displaying tags if the label is null/undefined

ryelle committed:
Style table header with compare button, search placeholder

LevinMedia commented:
@ryelle this is testing well for me, awesome work! 

A few comments: 

When creating a comparison, either through the card (#364) or using these checkboxes, when you click compare, the tabular data should be filtered to just what's being compared. 

As an example: someone might select two products from the tabular data and hit compare. It filters the entire report by the two products being compared, and it renders the comparison in the graph. The totals in the summary numbers should reflect the totals for the two products being compared. 

If they want to add a third product to the comparison, they can search for it in the comparison card, and update the comparison. When the comparison is updated, the new product is added to the graph and tabular data, and the totals update to reflect all three products in the comparison.

I also reviewed the original design documentation for this interaction, and noticed there was no way to quickly un-compare everything without unchecking them one by one, then clicking compare again - so I've added a small "Clear All" to the right of the compare button in the filter card. 

<img width="728" alt="screen shot 2018-09-11 at 11 49 05 am" src="https://user-images.githubusercontent.com/4500952/45381780-2b9cac80-b5bb-11e8-88b2-d379d6dc2895.png">

I've also updated the tag to use the same pattern from Gutenberg. I'll open up a new issue for that since it's used in a few different components. 

For reference - the changes noted above are reflected at the component master thread: p6riRB-3lz-p2

psealock commented on client/components/table/index.js:
```diff
@@ -151,6 +258,10 @@ TableCard.propTypes = {
 		} )
 	),
+	/**
+	 * A list of IDs, matching to the row list so that id[ 0 ] contains the object ID for the object displayed in row[ 0 ].
```
`id[ 0 ]` -> `ids[ 0 ]`

psealock commented on client/components/table/index.js:
```diff
@@ -67,33 +93,111 @@ class TableCard extends Component {
+					compareBy && (
+						<Button onClick={ this.onCompare } isDefault>
+							{ __( 'Compare', 'wc-admin' ) }
+						</Button>
```
Disable button when only one checkmark is checked or the selected rows are already being compared?

psealock replied:
Just looking at the designs and I see it might be something to leave for later

ryelle replied:
Yeah, this still needs to be done on the Compare filter itself too #364 - I figured that could be done with a future PR

psealock commented on client/components/filters/compare/index.js:
```diff
@@ -31,12 +32,23 @@ class CompareFilter extends Component {
+
+		const prevIds = getIdsFromQuery( prevQuery[ param ] );
+		const currentIds = getIdsFromQuery( query[ param ] );
+		if ( ! isEqual( prevIds.sort(), currentIds.sort() ) ) {
```
Can you compare `state.selected` against `currentIds` to see if they are the same and return without setting state? The reason is manipulating just the Compare card causes the tags to be fetched and disappear

psealock reviewed (commented):
works well, I just have a few comments

ryelle committed:
Prevent setting just an ID list as the state, as this will wipe out already-fetched tag labels

ryelle committed:
Update docs

ryelle committed:
Shortcut out of fetching tag labels if the query is empty

ryelle commented:
@LevinMedia thanks for the design review & behavior notes– the data interactions (showing just the compared products data) will have to wait for real API calls. I've also added the "clear all" button as a task to #364. Did you open the new issue for the token/tag component? 

@psealock I've addressed your feedback - rather than setting the selected state to the IDs while fetching the labels (which causes the existing labels to disappear), we just leave them as-is & update once the real content is loaded in.

LevinMedia commented:
@ryelle 

>Did you open the new issue for the token/tag component?

Just now. Thank you! https://github.com/woocommerce/wc-admin/issues/423

psealock reviewed (approved):
Code looks good, and behaviour is as expected
